File: 0242. Valid Anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # O(n) time complexity
        # O(m + n) space complexity

        if len(s) != len(t) or set(s) != set(t):
            return False

        sDict, tDict = {}, {}
        for sChar, tChar in zip(s, t):
            sDict[sChar] = sDict.get(sChar, 0) + 1
            tDict[tChar] = tDict.get(tChar, 0) + 1

        if sDict != tDict:
            return False
        return True

File: 0242. Valid Anagram.cpp
class Solution {
public:
    // O(n + m) time and O(max(m, n)) space complexity
    bool isAnagram(string s, string t) {
        unordered_map<char, int> charMap;

        for (char &ch : s) {
            charMap[ch]++;
        };

        for (char &ch : t) {
            if (charMap[ch] == 1) {
                charMap.erase(ch);
                continue;
            };
            charMap[ch]--;
        };

        return charMap.empty();
    };
};
